### ISSUE TITLE
Suggestion: Convenience functions for common scenarios such as Font Awesome icons

### ISSUE THREAD

**Issue by rmunn** [MangelMaxime/Fable.Elmish.Bulma#24]
Looking at the [`Icon` documentation](https://mangelmaxime.github.io/Fable.Elmish.Bulma/#elements/icon), I was struck by the verbosity of the resulting code, and how much is repeated. While I wouldn't want `Icon.icon` changed since there are times when you need to specify things explicitly, I wonder if there could be a convenience function for much of it. E.g.,:

```fsharp
let faIcon styles name =
    Icon.icon styles [ i [ ClassName (sprintf "fa fa-%s" name) ] [ ] ]
```

Then the icon sample code would go from this:

```fsharp
Icon.icon [ Icon.isSmall ]
    [ i [ ClassName "fa fa-home" ] [ ] ]
Icon.icon [ ]
    [ i [ ClassName "fa fa-home" ] [ ] ]
Icon.icon [ Icon.isMedium ]
    [ i [ ClassName "fa fa-home" ] [ ] ]
Icon.icon [ Icon.isLarge ]
    [ i [ ClassName "fa fa-home" ] [ ] ]
```

to this:

```fsharp
Icon.faIcon [ Icon.isSmall ] "home"
Icon.faIcon [ ] "home"
Icon.faIcon [ Icon.isMedium ] "home"
Icon.faIcon [ Icon.isLarge ] "home"
```

Similar functions for other icon sets would be easy enough to write, too. E.g., for [Typicons](http://www.typicons.com/more/how-to-use-typicons/) the function would look like:

```fsharp
let typIcon styles name =
    Icon.icon styles [ span [ ClassName (sprintf "typcn typcn-%s" name) ] [ ] ]
```

Such convenience functions would be easy enough to create in user code, of course. But if everyone's going to be creating a bunch of similar functions (and I think they will), then it might be nice to include such convenience functions in the library.

Now I'll play "devil's advocate" for a bit and argue against my own proposal.

One counterargument would be "If you get used to the convenience function, you'll forget how to use the basic function when you need to add something like an onClick handler that the convenience function doesn't provide." But at least in the specific case of icons, I don't think that's likely: most people who would need an onClick handler would be putting it on the larger element that include the icon, not on the icon itself.

The other counterargument, and probably the best one, would be, "It doesn't match the Fable.Elmish.Bulma code style." I.e., the rest of what you're building in Elmish.Bulma has an *obvious* 1-to-1 mapping to Bulma's CSS styles, and convenience functions like the `faIcon` function I'm suggesting are out of scope because they don't have an *obvious* mapping. That would mean that these are best left to users to define for themselves, rather than putting them in the library. I could easily find this argument persuasive, in which case it might be best to leave a `faIcon` function out of the Fable.Elmish.Bulma library after all. But I believe the idea is at least worth considering, even if the ultimate decision is not to include it.

**Comment by MangelMaxime** [comment]
Hello @rmunn , I have been asking my self the same question several times.

I agree with you that for the moment I am focusing on porting a 1-to-1 mapping to Bulma's CSS styles.
I am also always following the "React DSL" style:

1. An element name
2. A list of properties
3. A list of children

Eg: `div [  ] [  ]`

The reasons are simple, when I first start writing Elmish.Bulma I was making a lot of concession and adapting the function/DSL depending on their usage and even I was lost most of the time because I wasn't following the same syntax every time. So I decided to rework the whole lib to always follow the "React DSL" at first and later when I have more experience with it decide if compromise like the one you propose can be done.

All to say that the 1-to-1 mapping with Bulma's CSS style will always follow this convention as it make easy for users to use Elmish.Bulma in their project. 
However, I agree with you with the need of convenience functions where it make senses: **Icons are definitely a good place for it**

I have some ideas on how to include Font-Awesome support into Elmish.Bulma, I will create a PR so we can discuss it from there.
For others icon sets, we will be able to follow the sample convention.

**Comment by alfonsogarciacaro** [comment]
I completely agree with both points of view: The API of Elmish.Bulma may feel a bit verbose at times but this is better for maintenance (maybe updates with the Bulma framework can be automated somehow), compatibility with other React code, etc. But at the same time, it's nice to have helpers for commonly used patterns like Font Awesome icons. Probably we can keep the best of both worlds by leaving the basic API untouched and putting the helpers in a special namespace like `Helpers` or `Extra`... as @MangelMaxime just did in his PR ;)

One quick note about the helpers, in most cases if you just make the helper function `inline` you can save most of the overhead in the generated JS code. Also, I would avoid things like `sprintf` (currently Fable doesn't optimize it) when possible and just use the add operator. Actually, this could be a good time to optimize int and string literal operations when compiling, which is a long-time requested feature in Fable :wink:

**Comment by MangelMaxime** [comment]
@alfonsogarciacaro About the inline you speak about this function: https://github.com/MangelMaxime/Fable.Elmish.Bulma/pull/25/files#diff-ab2f5bc01fe44873d6c672ea0a0426eeR816 ?

So I should replace all my `sprintf` with `+` operator into Elmish.Bulma ?

**Comment by alfonsogarciacaro** [comment]
@MangelMaxime Yes, it will help to make that function inline (so the generated JS code is basically the same as when using the normal API). About the other question, if you're just using `sprintf` to concatenate strings (`sprintf "Hello %s!" "World"`), yes, it's more performant (don't have exact numbers) to just use `+`. Also in .NET F# unless the compiler makes any optimization I'm not aware of.

**Comment by MangelMaxime** [comment]
@alfonsogarciacaro Ok thanks for the info and seems logic in fact.
Don't know why but since I learn about `sprintf` I am abusing of it :)

**Comment by MangelMaxime** [comment]
The requested feature is now available in version [0.1.0-alpha9](https://github.com/MangelMaxime/Fable.Elmish.Bulma/blob/master/RELEASE_NOTES.md#010-alpha9)
